packages/editor/src/store/selectors.js

```diff
@@ -66,6 +67,21 @@ const ONE_MINUTE_IN_MS = 60 * 1000;
  */
 const EMPTY_ARRAY = [];
 
+/**
+ * Given a block type object, returns true if an attribute value is sourced
+ * from a meta property, or false otherwise.
+ *
+ * The value is cached weakly by strict object reference to the provided block
+ * type argument.
+ *
+ * @param {WPBlockType} blockType Block type against which to test.
+ *
+ * @return {boolean} Whether block type has an meta-sourced attribute.
+ */
+const hasMetaSourcedAttribute = createWeakCache( ( blockType ) => {
```

Comment:
Can‘t we just use memoize instead of rolling our own? We already use it elsewhere, e.g. the i18n package.

Reply:
There's a small but important difference in how this works using [`WeakMap`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap) as the cache storage, where normal memoization would hold on to values in memory forever, this would allow garbage collection to occur.

>The second inconvenience is a memory leak because the arrays ensure that references to each key and each value are maintained indefinitely. These references prevent the keys from being garbage collected, even if there are no other references to the object.  This would also prevent the corresponding values from being garbage collected.

```js
memoized = memoize( _.stubTrue );
weakCached = createWeakCache( _.stubTrue );

var obj = { a: 1 };
weakCached( obj );
// [ { a: 1 } => true ]
memoized( obj );
// [ { a: 1 } => true ]

obj = { b: 2 };

weakCached( obj );
// [ { b: 2 } => true ]
memoized( obj );
// [ { a: 1 } => true, { b: 2 } => true ]
//   ┃
//   ┗━ This will be stuck in memory forever, despite there being no other
//      reference to the object.
```

For blocks, we'd really only achieve some significant benefit if we expect blocks to be overwritten at any point. Practically speaking this probably wouldn't occur, so we could potentially be fine with a standard memoization, albeit with a very fragile guarantee against memory leaks.

`rememo` (`createSelector`) has some internal tricks to achieve a similar effect, so there might be some options there, but it'd be quite indirect.

Reply:
Lodash's documentation includes a note about swapping its underlying cache behavior, though as described it appears to apply globally. Each memoized function has its own cache, so I'd wonder if it might be possible to assign to it directly.

https://lodash.com/docs/4.17.11#memoize